### PR TITLE
Feat: Support for `--bprint` in language server.

### PR DESCRIPTION
**Pull request by MikaelMayer**
I add this as a way to simplify boogie debugging, especially in cases when VSCode is not consistent with the command-line like in #4205

I did not add any test for that as it's meant for Dafny development only, but I tested it manually. I borrowed the code from the driver.
Should I still list it as a feature?

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs [hunk 1]
```diff
@@ -55,6 +56,12 @@ DafnyOptions options
       return translated.SelectMany(t => {
         var (_, boogieProgram) = t;
         var results = engine.GetImplementationTasks(boogieProgram);
+        if (engine.Options.PrintFile != null) {
```
Comment (keyboardDrummer) [inline Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs:59]: I would put this in a separate traversal outside of the SelectMany.

```c#
if (engine.Options.PrintFile) {
  var nmodules = Translator.VerifiableModules(program).Count();
  foreach(var boogieProgram in translated) {
    ..
  }
}
```

File: Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs [hunk 2]
```diff
@@ -42,6 +42,7 @@ DafnyOptions options
       CancellationToken cancellationToken) {
       var program = document.Program;
       var errorReporter = (DiagnosticErrorReporter)program.Reporter;
+      var nmodules = Translator.VerifiableModules(program).Count();
```
Comment (keyboardDrummer) [inline Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs:45]: Can you give the introduced variables fuller names?

File: Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs [hunk 3]
```diff
@@ -52,10 +52,17 @@ DafnyOptions options
 
       cancellationToken.ThrowIfCancellationRequested();
 
+      if (engine.Options.PrintFile != null) {
+        var moduleCount = Translator.VerifiableModules(program).Count();
+        foreach (var t in translated) {
```
Comment (keyboardDrummer) [inline Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs:57]: You can immediately deconstruct the tuple here, so you can better names than `t` and `Item1` and `Item2`

**Review by keyboardDrummer — DISMISSED** [review]

File: Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs [hunk 4]
```diff
@@ -52,10 +52,17 @@ DafnyOptions options
...
+          var fileName = moduleCount > 1 ? DafnyMain.BoogieProgramSuffix(engine.Options.PrintFile, t.Item1) : engine.Options.PrintFile;
+          ExecutionEngine.PrintBplFile(engine.Options, fileName, t.Item2, false, false, engine.Options.PrettyPrint);
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs:59]:
```suggestion
        foreach (var (suffix, program) in translated) {
          var fileName = moduleCount > 1 ? DafnyMain.BoogieProgramSuffix(engine.Options.PrintFile, suffix) : engine.Options.PrintFile;
          ExecutionEngine.PrintBplFile(engine.Options, fileName, program, false, false, engine.Options.PrettyPrint);
```

File: Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs [hunk 5]
```diff
@@ -52,10 +52,17 @@ DafnyOptions options
 
       cancellationToken.ThrowIfCancellationRequested();
 
+      if (engine.Options.PrintFile != null) {
+        var moduleCount = Translator.VerifiableModules(program).Count();
+        foreach (var (suffix, program) in translated) {
+          var fileName = moduleCount > 1 ? DafnyMain.BoogieProgramSuffix(engine.Options.PrintFile, suffix) : engine.Options.PrintFile;
+          ExecutionEngine.PrintBplFile(engine.Options, fileName, program, false, false, engine.Options.PrettyPrint);
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer/Language/DafnyProgramVerifier.cs:59]:
```suggestion
        foreach (var (suffix, boogieProgram) in translated) {
          var fileName = moduleCount > 1 ? DafnyMain.BoogieProgramSuffix(engine.Options.PrintFile, suffix) : engine.Options.PrintFile;
          ExecutionEngine.PrintBplFile(engine.Options, fileName, boogieProgram, false, false, engine.Options.PrettyPrint);
```

**Pull request merged**